### PR TITLE
fix(xgboost): Remove xgboost dependency to reduce image sizes

### PR DESCRIPTION
**Pull request by vimalk78**
`xgboost` and its dependency `nvidia` each contribute to half the size of the model-server base image.

  This commit:
  - removes `xgboost`, and its dependency nvidia from the project
  - changes to model training code which uses `xgboost` to detect if `xgboost` is not present and skips execution

Image size comparison:

  - Image sizes before change
  ```
  quay.io/sustainable_computing_io/kepler_model_server        v0.7.11-2        dce5c5495a67   18 minutes ago   1.21GB
  quay.io/sustainable_computing_io/kepler_model_server_base   v0.7.11-2        c59e2a1a5f43   27 minutes ago   1.19GB
  ```

  - Image sizes after change
  ```
  quay.io/sustainable_computing_io/kepler_model_server        v0.7.11-2        1521f5b2612d   2 hours ago      688MB
  quay.io/sustainable_computing_io/kepler_model_server_base   v0.7.11-2        5ae7b07a679b   2 hours ago      686MB
  ```

 Any time if we need to use xgboost, just add xgboost to `pyproject.toml`, and re-run the test and pipeline.

# Checklist for PR Author

---

In addition to approval, the author must confirm the following check list:

- [x] Run the following command to format your code:

  ```bash
  make fmt
  ```

- [x] Create issues for unresolved comments and link them to this PR. Use one of the following labels:
  - `must-fix`: The logic appears incorrect and must be addressed.
  - `minor`: Typos, minor issues, or potential refactoring for better readability.
  - `nit`: Trivial issues like extra spaces, commas, etc.

### CODE REVIEW

**Comment by vimalk78** [comment]
Cc: @sunya-ch

File: src/kepler_model/estimate/model/model.py [hunk 1]
```diff
@@ -5,7 +5,15 @@
 
 from kepler_model.estimate.model.curvefit_model import CurveFitModelEstimator
 from kepler_model.estimate.model.scikit_model import ScikitModelEstimator
-from kepler_model.estimate.model.xgboost_model import XgboostModelEstimator
+
+have_xgboost = True
```
Comment (sthaha) [inline src/kepler_model/estimate/model/model.py:9]: w
```suggestion
import importlib.util 

has_xgboost = importlib.util.find_spec("xgboost") is not None
```

File: src/kepler_model/estimate/model/model.py [hunk 2]
```diff
@@ -5,7 +5,15 @@
...
+try:
+    import xgboost
+except ImportError:
+    have_xgboost = False
```
Comment (sthaha) [inline src/kepler_model/estimate/model/model.py:13]:
```suggestion
import importlib.util 

HAS_XGBOOST = importlib.util.find_spec("xgboost") is not None
```

also consider using python convention of naming constants in UPPERCASE

File: src/kepler_model/util/train_types.py [hunk 3]
```diff
@@ -45,8 +45,15 @@
...
     "SVRRegressorTrainer",
-    "XgboostFitTrainer",
 ]
+have_xgboost = True
+try:
+    import xgboost
+except:
+    have_xgboost = False
+if have_xgboost:
+    no_weight_trainers.append("XgboostFitTrainer")
```
Comment (sthaha) [inline src/kepler_model/util/train_types.py:55]:
```suggestion
try:
    import xgboost
    no_weight_trainers.append("XgboostFitTrainer")
except ImportError:
    pass
```

File: src/kepler_model/estimate/model/model.py (hunk 2 above)
Comment (vimalk78) [reply src/kepler_model/estimate/model/model.py:13]: any advantage of using importlib than  importing the package directly?

Comment (sthaha) [reply src/kepler_model/estimate/model/model.py:13]: yes, we don't want to import xgboost  (unused import)

File: src/kepler_model/util/train_types.py (hunk 3 above)
Comment (sthaha) [reply src/kepler_model/util/train_types.py:55]: On second thought, we don't want import xgboost rather only need to know if xgboost can be imported, hence could you please use `importlib.unit.find_spec` ?

File: src/kepler_model/estimate/model/model.py (hunk 2 above)
Comment (vimalk78) [reply src/kepler_model/estimate/model/model.py:13]: actually we want to keep the import for the case when we want to use the xgboost trainers. only change needed is to add back xgboost dependency in pyproject.toml

@KaiyiLiu1234 @sunya-ch please comment here.

**Comment by vimalk78** [comment]
> @vimalk78 could you please ensure the tests run by using the `base.changed` = false hack ?

instead, i have removed the `base_changed`  `if` condition check in the tekton and integration tests. Since these tests need to be run on every PR, and these tests need to detect `base_changed` only to change the base image while building model-server.

**Comment by vimalk78** [comment]
please rerun the tests

**Review by sthaha — COMMENTED** [review]
LGTM

**Comment by vimalk78** [comment]
Cc: @sunya-ch @rootfs

**Comment by sthaha** [comment]
@sunya-ch, I am going to merge this in the interest of time. Lets fix any issue we find with jobs in different PR.
